services/s3/itest/s3_test.go

```diff
@@ -793,3 +793,175 @@ func TestHead(t *testing.T) {
+	upload1, uploadErr1 := client.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
+		Bucket:                  &bucket,
+		Key:                     &key,
+		ServerSideEncryption:    types.ServerSideEncryptionAwsKms,
+		SSEKMSKeyId:             &kmsKey,
+		SSEKMSEncryptionContext: &kmsContext,
+		Tagging:                 aws.String("foo=bar"),
+	})
+	if uploadErr1 != nil {
+		t.Fatal(uploadErr1)
+	}
+	uploadId1 := upload1.UploadId
+
+	var parts []types.CompletedPart
+	for i, s := range []string{"hello", " world"} {
+		output, err := client.UploadPart(ctx, &s3.UploadPartInput{
+			PartNumber: aws.Int32(int32(i)),
+			Bucket:     &bucket,
+			Key:        &key,
+			UploadId:   uploadId1,
+			Body:       strings.NewReader(s),
+		})
+		if output.ServerSideEncryption != types.ServerSideEncryptionAwsKms {
+			t.Fatal("missing SSE header")
+		}
+		if *output.SSEKMSKeyId != kmsKey {
+			t.Fatal("missing KMS key header: ", *output.SSEKMSKeyId)
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+		parts = append(parts, types.CompletedPart{
+			ETag:       output.ETag,
+			PartNumber: aws.Int32(int32(i)),
+		})
+	}
+
+	// Create a second multipart upload for test-key and upload some parts to it
+	upload2, uploadErr2 := client.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
+		Bucket:                  &bucket,
+		Key:                     &key,
+		ServerSideEncryption:    types.ServerSideEncryptionAwsKms,
+		SSEKMSKeyId:             &kmsKey,
+		SSEKMSEncryptionContext: &kmsContext,
+		Tagging:                 aws.String("foo=bar"),
+	})
+	if uploadErr2 != nil {
+		t.Fatal(uploadErr2)
+	}
+	uploadId2 := upload2.UploadId
+
+	var parts2 []types.CompletedPart
+	for i, s := range []string{"hello", " world"} {
+		output, err := client.UploadPart(ctx, &s3.UploadPartInput{
+			PartNumber: aws.Int32(int32(i)),
+			Bucket:     &bucket,
+			Key:        &key,
+			UploadId:   uploadId2,
+			Body:       strings.NewReader(s),
+		})
+		if output.ServerSideEncryption != types.ServerSideEncryptionAwsKms {
+			t.Fatal("missing SSE header")
+		}
+		if *output.SSEKMSKeyId != kmsKey {
+			t.Fatal("missing KMS key header: ", *output.SSEKMSKeyId)
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+		parts2 = append(parts, types.CompletedPart{
+			ETag:       output.ETag,
+			PartNumber: aws.Int32(int32(i)),
+		})
+	}
+
+	// Complete the multipart upload for the first upload
+	_, completeErr1 := client.CompleteMultipartUpload(ctx, &s3.CompleteMultipartUploadInput{
+		Bucket:   &bucket,
+		Key:      &key,
+		UploadId: uploadId1,
+		MultipartUpload: &types.CompletedMultipartUpload{
+			Parts: parts,
+		},
+	})
+
+	if completeErr1 != nil {
+		t.Fatal(completeErr1)
+	}
+
+	// Expect the second upload to fail completing since an object already exists with that key
+	_, completeErr2 := client.CompleteMultipartUpload(ctx, &s3.CompleteMultipartUploadInput{
+		Bucket:   &bucket,
+		Key:      &key,
+		UploadId: uploadId1,
```

Comment:
ah, you are passing uploadId1 instead of uploadId2 here, which results in a different error (since the upload got compelted and the id no longer exists).

I'm going to fix up this test by doing the upload twice in a loop, that will make it clearer what's going on